test/instrument/control/if.t

```diff
@@ -97,3 +97,116 @@ tail position iff the whole if-expression is in tail position.
     else (
       ___bisect_visit___ 4;
       print_endline "bar")
+
+When the if-then-else is in a sequence, the next-expressions are
+instrumented as expected.
+
+  $ bash ../test.sh <<'EOF'
+  > let _ =
+  >   if bool_of_string "true" then print_endline "foo" else print_endline "bar";
+  >   print_endline "this";
+  >   print_endline "that";
+  > EOF
+  let _ =
+    if bool_of_string "true" then (
+      ___bisect_visit___ 5;
+      ___bisect_post_visit___ 2 (print_endline "foo"))
+    else (
+      ___bisect_visit___ 4;
+      ___bisect_post_visit___ 3 (print_endline "bar"));
+    ___bisect_post_visit___ 1 (print_endline "this");
+    ___bisect_post_visit___ 0 (print_endline "that")
+
+Where there's a raise or a failwith guarded by a conditional, the
+branch itself is instrumented, but the raising expression is not
+post-instrumented...
+
+  $ bash ../test.sh <<'EOF'
+  > let _f ~cond1 ~cond2 =
+  >   if cond1 then failwith "this";
+  >   if cond2 then raise Not_found;
+  >   print_endline "this";
+  >   print_endline "that";
+  > EOF
+  let _f ~cond1 ~cond2 =
+    ___bisect_visit___ 5;
+    if cond1 then (
+      ___bisect_visit___ 4;
+      failwith "this");
+    ___bisect_visit___ 3;
+    if cond2 then (
+      ___bisect_visit___ 2;
+      raise Not_found);
+    ___bisect_visit___ 1;
+    ___bisect_post_visit___ 0 (print_endline "this");
+    print_endline "that"
+
+... however note that this doesn't hold for all raising expressions.
+For example, if you use `invalid_arg`, or a raising helper, its
+application will be post-instrumented (unless all of its arguments are
+labeled !?), resulting in a code location that can never be visited,
+by design. These are currently acknowledged limitations.
```

Comment:
This is an acknowledged and inevitable limitation because there is no way for Bisect to know that a specific identifier refers to a function that always or never raises. In fact, it doesn't know that for the identifiers `failwith` or `invalid_arg` either, as those can be redefined and shadowed by the user, including by another module or a third-party library that is not even processed by Bisect during compilation of the user's particular project. I don't see that we need a test for this -- the actual behavior is covered by the normal `apply` tests.

Reply:
(Raises or diverges)

Reply:
> there is no way for Bisect to know that a specific identifier refers to a
> function that always or never raises

I agree. It's fair to say that "a strategy to limit non-visitable points, which
relies partly on identifiers, will inevitably encounter these limitations."
However, if we were to discover an alternative approach (assuming one exists),
we might still see improvements in this test. In other words, the insertion of
post-instrumentation might not be inevitable when considering the range of
possible instrumentation strategies.

Reply:
> (Raises or diverges)

Btw, I didn't understand the part where you said  "or diverges".

Reply:
> > (Raises or diverges)
> 
> Btw, I didn't understand the part where you said "or diverges".

Post-instrumentation checks that a function evaluates to a value. The two ways that it doesn't in OCaml are when it raises an exception or when it diverges.

Reply:
Those are the two ways that the out-edge of a`Pexp_apply` may *not* be followed.

Reply:
Does "diverging" encompass non-termination and potentially the effects in OCaml 5? Something else?

Regarding the effects, I'm optimistic that their influence on the strategy will
be similar to that of exceptions. I hope this won't necessitate too big of an
exploration of the OCaml 5 runtime :exploding_head:. Thank you again for your explanation!

Reply:
Yes, "diverging" is non-termination. OCaml 5 effects are supposed to always be returned from exactly once, so they should not affect control flow as visible to Bisect, although this is not enforced, and for effects that correspond to async computation the "exactly once" might be later than program termination, so in that sense they behave like exceptions.